### PR TITLE
Adds skip-breakpoints feature

### PR DESCRIPTION
**Pull request by iwek7**
Closes part of #27028
This implements feature of disabling and enabling breakpoints. I've added button next to debug icons:
Breakpoints enabled:
image

Breakpoints disabled:
image

This setting is stored in memory and is set to enabling breakpoints at every editor startup. Skipping breakpoints happens in remote application where flag value is checked. I've added new startup argument for app launched from godot as well as wired up skip-breakpoints button to peer stream communication with remote debugger in case of changes when game is already running.

Feedback very welcome, this is feature draft and can be changed - icons suggestions especially welcome (I sux at drawing). Also I'm not sure if I should do anything with local debugger.

Scenarios tested:

1. Turn on skip breakpoints -> run game -> observe breakpoints not working
2. Don't turn on skipbreakpoints -> run game -> observe breakpoints working
3. Run game with skip breakpoints off -> turn feature on -> observe breakpoints not working
4. Run game with skip breakpoints on -> turn feature off -> observe breakpoints working
5. Run game and stop in breakpoint -> turn skip breakpoints on -> observe breakpoints not working after clicking continue

### CODE REVIEW

File: editor/editor_run.h [hunk 1]
```diff
@@ -51,7 +51,7 @@ class EditorRun {
 
 public:
 	Status get_status() const;
-	Error run(const String &p_scene, const String &p_custom_args, const List<String> &p_breakpoints);
+	Error run(const String &p_scene, const String &p_custom_args, const List<String> &p_breakpoints, const bool &skip_breakpoints);
```
Comment (Geequlim) [inline editor/editor_run.h]: How about give `skip_breakpoints`  as default value as  `bool p_skip_breakpoints = false`.
In this way it should not break anything.

File: core/script_debugger_remote.h [hunk 2]
```diff
@@ -176,6 +178,8 @@ class ScriptDebuggerRemote : public ScriptDebugger {
 	virtual void profiling_end();
 	virtual void profiling_set_frame_times(float p_frame_time, float p_idle_time, float p_physics_time, float p_physics_frame_time);
 
+	virtual void set_skip_breakpoints(bool skip_breakpoints);
```
Comment (Geequlim) [inline core/script_debugger_remote.h]: Add `p_` prefix to the parameter name here.

**Review by Geequlim — CHANGES_REQUESTED** [review]

**Comment by Geequlim** [comment]
Not got time to compile and test.
The way to implement this feature looks good to me.

I am a bit worried about this situation.
When an error triggered at runtime without the breakpoints enabled.
```gdscript
func on_button_click():
    if self.get('no_such_property').no_such_value:
        breakpoint
    pass
```

File: editor/editor_run.h (hunk 1 above)
Comment (iwek7) [reply editor/editor_run.h]: And is it breaking anything? It seems to be called only in one place.

File: core/script_debugger_remote.h (hunk 2 above)
Comment (iwek7) [reply core/script_debugger_remote.h]: fixed

**Comment by iwek7** [comment]
Thanks a lot for review. I am not so sure what is the situation you are worried about, what do you mean by error triggered at runtime? Both in-code and normal breakpoints are disabled by this feature.

**Comment by Geequlim** [comment]
@iwek7  I mean does crash breakpoint work when debugging games with the breakpoint is disabled

**Comment by iwek7** [comment]
Not sure what crash breakpoint is. Do you mean just breakpoint keyword used in script? If breakpoints are disabled then they are not working as well.

**Comment by Geequlim** [comment]
If a runtime error occurs, the debugger will jump to the line of code that went wrong.
Will this behavior fail because the breakpoint is disabled?

**Comment by iwek7** [comment]
I somehow did not think about this feature. Indeed my code turns off crash breakpoints as well. Would we want crash breakpoints to work despite breakpoints being turned off? I think yes so I will try to fix that.

Thanks for paying attention.

**Comment by iwek7** [comment]
I will have time to work on this sometime next week

File: editor/editor_run.h (hunk 1 above)
Comment (iwek7) [reply editor/editor_run.h]: fixed anyway

**Comment by iwek7** [comment]
@Geequlim I've added code that prevents disabling breakpoints from working when error breakpoints are involved (changes are in separate commit for now). I've tested it for gdscript and it works. Sadly the way it is done is far away from satisfactory in my opinion: I had to extend debug function on script_language class with new optional argument that informs if breakpoint pause is caused by error. Sadly it is only easy to tell if breakpoint is error is from gdscript. From calls of this function from visual script it seems that it is called only in error case and there are no case for c# in godot code... . I have no real knowledge of c# and visual script in godot so it's super hard to tell what is going on.

I kinda started to doubt if this feature with changes to script_language is acceptable, we could take first version of it and accept that error breakpoints will be skipped too but I don't like it either...

File: drivers/gles2/shader_compiler_gles2.cpp [hunk 3]
```diff
@@ -758,11 +758,13 @@ String ShaderCompilerGLES2::_dump_node_code(SL::Node *p_node, int p_level, Gener
 				} break;
 
 				case SL::OP_SELECT_IF: {
+					code += "(";
```
Comment (vnen) [inline drivers/gles2/shader_compiler_gles2.cpp]: Those changes in shader compiler are unrelated to this feature. Please put these in another PR.

File: core/script_debugger_remote.h (hunk 2 above)
Comment (vnen) [reply core/script_debugger_remote.h]: This is not fixed yet.

File: core/script_debugger_remote.cpp [hunk 4]
```diff
@@ -1102,6 +1109,10 @@ void ScriptDebuggerRemote::profiling_set_frame_times(float p_frame_time, float p
 	physics_frame_time = p_physics_frame_time;
 }
 
+void ScriptDebuggerRemote::set_skip_breakpoints(bool skip_breakpoints) {
```
Comment (akien-mga) [inline core/script_debugger_remote.cpp]: Should be `p_skip_breakpoints` (thus removing the need for `this->` below).

File: core/script_debugger_remote.h (hunk 2 above)
Comment (akien-mga) [reply core/script_debugger_remote.h]: Not that occurrence :)

**Comment by vnen** [comment]
Changes look good apart from the couple of comments. Also needs a rebase to have all changes in a single commit.

File: drivers/gles2/shader_compiler_gles2.cpp (hunk 3 above)
Comment (iwek7) [reply drivers/gles2/shader_compiler_gles2.cpp]: It seems that I took some random commit in my branch by mistake, will fix

File: core/script_debugger_remote.cpp (hunk 4 above)
Comment (iwek7) [reply core/script_debugger_remote.cpp]: fixed

File: drivers/gles2/shader_compiler_gles2.cpp (hunk 3 above)
Comment (iwek7) [reply drivers/gles2/shader_compiler_gles2.cpp]: i dropped that commit, dunno what happend here

File: core/script_debugger_remote.h (hunk 2 above)
Comment (iwek7) [reply core/script_debugger_remote.h]: it is now :)

**Comment by iwek7** [comment]
I fixed issues and rebased, thanks for review

**Comment by akien-mga** [comment]
Thanks!

**Comment by chucklepie** [comment]
This feature is great as a temporary debug facility, but what it needs is an option to clear all breakpoints for those times when you have a load of them all over the project and just want to get rid of them.

image
